src/XCCDF_POLICY/xccdf_policy_remediate.c

```diff
@@ -464,8 +464,14 @@ static inline int _xccdf_fix_execute(struct xccdf_rule_result *rr, struct xccdf_
 				NULL
 			};
 
-			char *const envp[2] = {
+			char *oscap_bootc_build = getenv("OSCAP_BOOTC_BUILD");
+			char *oscap_bootc_build_kvarg = NULL;
+			if (oscap_bootc_build != NULL) {
+				oscap_bootc_build_kvarg = oscap_sprintf("OSCAP_BOOTC_BUILD=%s", oscap_bootc_build);
```

Comment:
This will leak memory no? We should call `free(oscap_bootc_build_kvarg)` once we are done using it.

Reply:
Good point! TBH I don't know how to do that because we need this memory to be allocated when the `execve` is called so we can't free it before, and we can't free it after either because the `execve` won't return (see the comment there). Also, I'm puzzled that the valgrind doesn't report this leak:

```
jcerny@fedora:~/work/git/openscap/build (pass_down)$ OSCAP_BOOTC_BUILD=YES ./run valgrind --leak-check=full --show-leak-kinds=all utils/oscap xccdf eval --remediate --results /tmp/x ../tests/API/XCCDF/unittests/test_oscap_bootc_pass_down.ds.xml
==63916== Memcheck, a memory error detector
==63916== Copyright (C) 2002-2024, and GNU GPL'd, by Julian Seward et al.
==63916== Using Valgrind-3.23.0 and LibVEX; rerun with -h for copyright info
==63916== Command: utils/oscap xccdf eval --remediate --results /tmp/x ../tests/API/XCCDF/unittests/test_oscap_bootc_pass_down.ds.xml
==63916== 
==63916== Thread 4 input_handler:
==63916== realloc() with size 0
==63916==    at 0x484B347: realloc (vg_replace_malloc.c:1800)
==63916==    by 0x49597D0: SEAP_packet_sexp2msg (seap-packet.c:209)
==63916==    by 0x495AE4B: SEAP_packet_recv (seap-packet.c:730)
==63916==    by 0x495C59A: SEAP_recvmsg (seap.c:360)
==63916==    by 0x494E3B3: probe_input_handler (input_handler.c:102)
==63916==    by 0x56A16D6: start_thread (pthread_create.c:447)
==63916==    by 0x5725413: clone (clone.S:100)
==63916==  Address 0x7f6dfe0 is 0 bytes after a block of size 0 alloc'd
==63916==    at 0x4843866: malloc (vg_replace_malloc.c:446)
==63916==    by 0x495928B: SEAP_packet_sexp2msg (seap-packet.c:110)
==63916==    by 0x495AE4B: SEAP_packet_recv (seap-packet.c:730)
==63916==    by 0x495C59A: SEAP_recvmsg (seap.c:360)
==63916==    by 0x494E3B3: probe_input_handler (input_handler.c:102)
==63916==    by 0x56A16D6: start_thread (pthread_create.c:447)
==63916==    by 0x5725413: clone (clone.S:100)
==63916== 
--- Starting Evaluation ---

Title   This rule always fails
Rule    xccdf_com.example.www_rule_test-fail
Result  fail


--- Starting Remediation ---

Title   This rule always fails
Rule    xccdf_com.example.www_rule_test-fail
Result  error

==63916== 
==63916== HEAP SUMMARY:
==63916==     in use at exit: 2,150 bytes in 17 blocks
==63916==   total heap usage: 298,764 allocs, 298,747 frees, 31,341,422 bytes allocated
==63916== 
==63916== Thread 1:
==63916== 27 bytes in 1 blocks are still reachable in loss record 1 of 7
==63916==    at 0x4843866: malloc (vg_replace_malloc.c:446)
==63916==    by 0x402631F: malloc (rtld-malloc.h:56)
==63916==    by 0x402631F: strdup (strdup.c:42)
==63916==    by 0x40151F8: _dl_load_cache_lookup (dl-cache.c:515)
==63916==    by 0x4009227: _dl_map_object (dl-load.c:2116)
==63916==    by 0x400CD7B: dl_open_worker_begin (dl-open.c:578)
==63916==    by 0x4001522: _dl_catch_exception (dl-catch.c:237)
==63916==    by 0x400C49F: dl_open_worker (dl-open.c:803)
==63916==    by 0x4001522: _dl_catch_exception (dl-catch.c:237)
==63916==    by 0x400C903: _dl_open (dl-open.c:905)
==63916==    by 0x5773654: do_dlopen (dl-libc.c:95)
==63916==    by 0x4001522: _dl_catch_exception (dl-catch.c:237)
==63916==    by 0x4001678: _dl_catch_error (dl-catch.c:256)
==63916== 
==63916== 27 bytes in 1 blocks are still reachable in loss record 2 of 7
==63916==    at 0x4843866: malloc (vg_replace_malloc.c:446)
==63916==    by 0x400C200: UnknownInlinedFun (rtld-malloc.h:56)
==63916==    by 0x400C200: _dl_new_object (dl-object.c:199)
==63916==    by 0x4007519: _dl_map_object_from_fd (dl-load.c:1053)
==63916==    by 0x4008FBF: _dl_map_object (dl-load.c:2249)
==63916==    by 0x400CD7B: dl_open_worker_begin (dl-open.c:578)
==63916==    by 0x4001522: _dl_catch_exception (dl-catch.c:237)
==63916==    by 0x400C49F: dl_open_worker (dl-open.c:803)
==63916==    by 0x4001522: _dl_catch_exception (dl-catch.c:237)
==63916==    by 0x400C903: _dl_open (dl-open.c:905)
==63916==    by 0x5773654: do_dlopen (dl-libc.c:95)
==63916==    by 0x4001522: _dl_catch_exception (dl-catch.c:237)
==63916==    by 0x4001678: _dl_catch_error (dl-catch.c:256)
==63916== 
==63916== 48 bytes in 6 blocks are still reachable in loss record 4 of 7
==63916==    at 0x4843866: malloc (vg_replace_malloc.c:446)
==63916==    by 0x546C5A0: do_malloc.constprop.0.isra.0 (global.c:1008)
==63916==    by 0x53EC0E4: UnknownInlinedFun (global.c:1030)
==63916==    by 0x53EC0E4: _gcry_xmalloc (global.c:1204)
==63916==    by 0x53ECE0E: UnknownInlinedFun (mpiutil.c:138)
==63916==    by 0x53ECE0E: UnknownInlinedFun (mpiutil.c:100)
==63916==    by 0x53ECE0E: UnknownInlinedFun (mpiutil.c:596)
==63916==    by 0x53ECE0E: UnknownInlinedFun (mpiutil.c:78)
==63916==    by 0x53ECE0E: global_init.part.0 (global.c:133)
==63916==    by 0x53ED1DC: UnknownInlinedFun (visibility.c:71)
==63916==    by 0x53ED1DC: UnknownInlinedFun (global.c:263)
==63916==    by 0x53ED1DC: gcry_check_version (visibility.c:70)
==63916==    by 0x4984ACE: crapi_init (crapi.c:43)
==63916==    by 0x48FD02F: oval_probe_session_libinit (oval_probe_session.c:101)
==63916==    by 0x56A6A4A: __pthread_once_slow.isra.0 (pthread_once.c:116)
==63916==    by 0x56A6AB8: pthread_once@@GLIBC_2.34 (pthread_once.c:143)
==63916==    by 0x48FD050: __init_once (oval_probe_session.c:116)
==63916==    by 0x48FD0C5: oval_probe_session_init (oval_probe_session.c:137)
==63916==    by 0x48FD184: oval_probe_session_new (oval_probe_session.c:157)
==63916== 
==63916== 144 bytes in 6 blocks are still reachable in loss record 5 of 7
==63916==    at 0x4843866: malloc (vg_replace_malloc.c:446)
==63916==    by 0x546C5A0: do_malloc.constprop.0.isra.0 (global.c:1008)
==63916==    by 0x53EC0E4: UnknownInlinedFun (global.c:1030)
==63916==    by 0x53EC0E4: _gcry_xmalloc (global.c:1204)
==63916==    by 0x53ECE01: UnknownInlinedFun (mpiutil.c:99)
==63916==    by 0x53ECE01: UnknownInlinedFun (mpiutil.c:596)
==63916==    by 0x53ECE01: UnknownInlinedFun (mpiutil.c:78)
==63916==    by 0x53ECE01: global_init.part.0 (global.c:133)
==63916==    by 0x53ED1DC: UnknownInlinedFun (visibility.c:71)
==63916==    by 0x53ED1DC: UnknownInlinedFun (global.c:263)
==63916==    by 0x53ED1DC: gcry_check_version (visibility.c:70)
==63916==    by 0x4984ACE: crapi_init (crapi.c:43)
==63916==    by 0x48FD02F: oval_probe_session_libinit (oval_probe_session.c:101)
==63916==    by 0x56A6A4A: __pthread_once_slow.isra.0 (pthread_once.c:116)
==63916==    by 0x56A6AB8: pthread_once@@GLIBC_2.34 (pthread_once.c:143)
==63916==    by 0x48FD050: __init_once (oval_probe_session.c:116)
==63916==    by 0x48FD0C5: oval_probe_session_init (oval_probe_session.c:137)
==63916==    by 0x48FD184: oval_probe_session_new (oval_probe_session.c:157)
==63916== 
==63916== 600 bytes in 1 blocks are still reachable in loss record 6 of 7
==63916==    at 0x484B133: calloc (vg_replace_malloc.c:1675)
==63916==    by 0x40143BF: UnknownInlinedFun (rtld-malloc.h:44)
==63916==    by 0x40143BF: _dl_check_map_versions (dl-version.c:280)
==63916==    by 0x400CE32: dl_open_worker_begin (dl-open.c:646)
==63916==    by 0x4001522: _dl_catch_exception (dl-catch.c:237)
==63916==    by 0x400C49F: dl_open_worker (dl-open.c:803)
==63916==    by 0x4001522: _dl_catch_exception (dl-catch.c:237)
==63916==    by 0x400C903: _dl_open (dl-open.c:905)
==63916==    by 0x5773654: do_dlopen (dl-libc.c:95)
==63916==    by 0x4001522: _dl_catch_exception (dl-catch.c:237)
==63916==    by 0x4001678: _dl_catch_error (dl-catch.c:256)
==63916==    by 0x57735C0: dlerror_run (dl-libc.c:45)
==63916==    by 0x577380F: __libc_dlopen_mode (dl-libc.c:162)
==63916== 
==63916== 1,260 bytes in 1 blocks are still reachable in loss record 7 of 7
==63916==    at 0x484B133: calloc (vg_replace_malloc.c:1675)
==63916==    by 0x400BF1D: UnknownInlinedFun (rtld-malloc.h:44)
==63916==    by 0x400BF1D: _dl_new_object (dl-object.c:92)
==63916==    by 0x4007519: _dl_map_object_from_fd (dl-load.c:1053)
==63916==    by 0x4008FBF: _dl_map_object (dl-load.c:2249)
==63916==    by 0x400CD7B: dl_open_worker_begin (dl-open.c:578)
==63916==    by 0x4001522: _dl_catch_exception (dl-catch.c:237)
==63916==    by 0x400C49F: dl_open_worker (dl-open.c:803)
==63916==    by 0x4001522: _dl_catch_exception (dl-catch.c:237)
==63916==    by 0x400C903: _dl_open (dl-open.c:905)
==63916==    by 0x5773654: do_dlopen (dl-libc.c:95)
==63916==    by 0x4001522: _dl_catch_exception (dl-catch.c:237)
==63916==    by 0x4001678: _dl_catch_error (dl-catch.c:256)
==63916== 
==63916== LEAK SUMMARY:
==63916==    definitely lost: 0 bytes in 0 blocks
==63916==    indirectly lost: 0 bytes in 0 blocks
==63916==      possibly lost: 0 bytes in 0 blocks
==63916==    still reachable: 2,106 bytes in 16 blocks
==63916==         suppressed: 44 bytes in 1 blocks
==63916== 
==63916== For lists of detected and suppressed errors, rerun with: -s
==63916== ERROR SUMMARY: 4 errors from 1 contexts (suppressed: 0 from 0)
```

Reply:
I see, so it doesn't need to be freed. I haven't realized that `execve` is called later on. `execve` replaces the current process memory with the new one (newly initialized stack, heap, and initialized and uninitialized data segments), so there is nothing to be freed once it is called.

Reply:
:scream: